jpype/dbapi2.py

```diff
@@ -0,0 +1,1444 @@
+
+                # Find and apply the adapter
+                a = cx._adapters.get(type(p), None)
+                if a is not None:
+                    p = a(p)
+
+                # Find the setter
+                if types[i] is None:
+                    s = cx._setters(cx, meta, i, type(p))
+                    types[i] = s
+                if s is None:
+                    raise _UnsupportedTypeError("no setter found for '%s'" % type(p).__name__)
+                s.set(self._statement, i + 1, p)
+            if count != i + 1:
+                raise ProgrammingError("incorrect number of parameters (%d!=%d)" % (count, i + 1))
+            # Cache
+            if hasattr(cx._setters, "_cachable"):
+                self._parameterTypes = types
+        else:
+            raise _UnsupportedTypeError("'%s' parameters not supported" % (type(params).__name__))  # pragma: no cover
+
+    def _onResultSet(self, rs):
+        meta = rs.getMetaData()
+        self._resultSet = rs
+        self._resultSetMeta = meta
+        self._resultSetCount = meta.getColumnCount()
+        self._columnTypes = None
+
+    def _fetchRow(self, converters):
+        cx = self._connection
+        count = self._resultSetCount
+        meta = self._resultSetMeta
+
+        byPosition = False
+        if converters is _default:
+            converters = cx._converters
+        if isinstance(converters, typing.Sequence):
+            if len(converters) != count:
+                raise ProgrammingError("converter list size incorrect")
+            byPosition = True
+
+        # Get all the column types
+        if self._columnTypes is None:
+            gk = cx._getters
+            # We need the type information for the columns
+            self._columnTypes = [gk(cx, meta, i) for i in range(count)]
+        if len(self._columnTypes) != count:
+            raise ProgrammingError("incorrect number of columns")
+        try:
+            row = []
+            for idx in range(count):
+                tp = self._columnTypes[idx]
+                # Fetch the value
+                value = tp.get(self._resultSet, idx + 1, False)
+                if value is None or converters is None:
+                    row.append(value)
+                elif byPosition:
+                    # find the column converter by type
+                    converter = converters[idx]
+                    row.append(converter(value))
+                else:
+                    # find the column converter by type
+                    converter = cx._converters.get(type(value), _nop)
+                    row.append(converter(value))
+            return row
+        except TypeError as ex:
+            raise _UnsupportedTypeError(str(ex)) from ex
+
+    def _validate(self):
+        """ Called before any method that requires the statement to be open. """
+        if self._closed or self._jcx.isClosed() or threading.get_ident() != self._thread:
```

Comment:
Please raise a different error in the mismatching thread case.

Reply:
Will do.  Errors are very poorly defined in the the dbapi2 spec.   It seems like to make dbapi2 generic to whatever the underlying implementation that just left all the errors as "an Error will be raised" but at the same time defined a whole tree of error types.   This leaves it hard to decide what error type applies to what situation.   I split the thread error to an Error with the message indicating it is a thread problem.  But which specific error it should be is poorly defined.  I tried to read through other dbapi2 wrappers to try to puzzle it out, but found no consistency at all.

Reply:
Sorry for the confusion, I'm just saying that the error message should be specific to the issue (threading).

Reply:
Ah seems like the threading issue is already firing up.